Signet

kallewoof opened the pull request:
Placeholder PR for reviewers.

wraithm commented on contrib/signet/issuer/issuer.sh:
```diff
@@ -0,0 +1,75 @@
+log "hit ^C to stop"
+
+while true; do
+    let rv=$RANDOM%$randinterval
```
Is this a uniform random variable in the interval? It might be interesting to generate times from an exponential distribution to more accurately simulate block inter-arrival times.

kallewoof replied:
That's a fair point. Not sure how easy it would be to do in bash though.

kallewoof replied:
I guess it could simply do a loop of "sleep 30; if random(probability*30/600)==0 then break" kind of thing.

instagibbs commented on src/validation.cpp:
```diff
@@ -935,9 +936,17 @@ bool ReadBlockFromDisk(CBlock& block, const FlatFilePos& pos, const Consensus::P
-    if (!CheckProofOfWork(block.GetHash(), block.nBits, consensusParams))
-        return error("ReadBlockFromDisk: Errors in block header at %s", pos.ToString());
+    // We skip POW checks for genesis block with signet
+    if (!g_signet_blocks || block.GetHash() != consensusParams.hashGenesisBlock) {
```
just unconditionally skip PoW checks for genesis block, this block is overly confusing :)

kallewoof replied:
Trying to avoid knee-jerk as this will run on mainnet as well. Maybe you're right tho.

instagibbs replied:
I think it's harder to read and be sure it's not messing with bitcoin logic. YMMV

instagibbs commented on src/signet.cpp:
```diff
@@ -0,0 +1,58 @@
+{
+    SimpleSignatureChecker bsc(signet_hash);
+    CScript solution = CScript(signature.begin(), signature.end());
+    return VerifyScript(solution, g_signet_blockscript, nullptr, STANDARD_SCRIPT_VERIFY_FLAGS, bsc);
```
I don't think you should use standard flags directly. These can be introduced and removed, at least the policy ones.

Instead use `GetBlockScriptFlags` or pass the flags in to get the currently enforced script flags.

kallewoof replied:
Good point!

instagibbs reviewed (commented):
maybe you should make the coinbase signature "commitment" be a serialized witness stack, and then you could support segwit signing(implicitly a blank scriptSig for block). This means you can support v0 witness, and any subsequent softfork as it's activated on signet.

This will make the coinbase tx a bit larger(witnessScript as last stack element for v0), but I think is more pleasing to work with.

This is exactly how I've written dynafed for Elements, and it means you can get things like taproot blocksigs without changing anything about signet itself.

JeremyRubin commented:
I don't get it -- why not just sign the find and deleted of signature data coinbase txn. This already contains all the leaves of all the txns :)

kallewoof commented:
@JeremyRubin 
> I don't get it -- why not just sign the find and deleted of signature data coinbase txn. This already contains all the leaves of all the txns :)

Witness commitment is not a signature, it's an output in the coinbase tx, so find and deleting sigs will not actually do it here (unless I move the signet sig into the coinbase input sigScript).

kallewoof commented:
@instagibbs Can you point me at code where you are doing this? I'm having trouble following your description.

JeremyRubin commented:
I think I was unclear. Here's a simple version for explanation:

Take the standard coinbase TX. Hash it. That hash is C.

Take the Blockheader minus merkleroot and nonce. Hash it. That hash is D.

Now compute H(C || D) = E.

Now sign E with PK and append (R,S) as an op_return on the CB txn.

When processing the block, remove the op_return, compute C, D, E, verify((R,S), E, PK)

kallewoof commented:
@JeremyRubin Thanks for the description! I understand what you propose, but I don't think I get what makes it an improvement. Also do note that signet right now is an arbitrary script, so (R, S) may not be sufficient in some cases (e.g. there may be some non-aggregated multisig or something). Just to be clear, right now:

Take the standard coinbase TX.

Set the signet commitment extension (in the segwit witness commitment) to {}.

Take the merkle root of the resulting TX and the remaining transactions in the block.

Sign H(prevblock | ... | merkleroot | ...).

In my case, I have to copy the coinbase and reset the signet commitment. In your case I have to remove the (R,S) op_return in the CB to get the hash. 

I.e. signing becomes:

1. Zero-out signet commitment.
2. Calc merkle root from zeroed-out coinbase + other txs.
3. Sighash = hash of this + cruft.
4. Sign sighash and set signet commitment to results

Verifying becomes:

1. Set sig to signet commitment and zero-out signet commitment.
2. Calc merkle root for zeroed-out coinbase + other txs.
3. Sighash = hash of this + cruft
4. Verify sig.

JeremyRubin commented:
Gotcha -- is this updated from before? The Merkle Root approach seems fine to me. Previously the CB txn was not committed.

instagibbs commented:
@kallewoof it's a lot of code, sorry, but my point was that you can have the script challenge just be interpreted like a scriptPubKey, then the given signature is included in a normal witness stack i.e., `CScriptWitness`: https://github.com/ElementsProject/elements/pull/642/files#diff-c8c12afab048e99c8478aee5b60bb6c0R62

This makes future softforks in bitcoin consensus "just work" with signet.

kallewoof commented:
@JeremyRubin 
> Gotcha -- is this updated from before? The Merkle Root approach seems fine to me. Previously the CB txn was not committed.

Yeah, sorry, I should have been clearer. I got feedback and tweaked things around (it now includes PoW, so software can literally hook into a signet without any modifications, if they are fine not checking the block signature).

@instagibbs Wait, so the signet signature would be an extra-push on the coinbase input signature? I like that idea.

instagibbs commented:
You can serialize it wherever you like, it's just a push of a vector of vector of bytes rather than a push of vector of bytes

kallewoof committed:
signet: hard-coded parameters for Signet Global Network IV (2020-03-05)

Running with -signet without any other parameters will use this network's parameters.

kallewoof commented:
Ping @practicalswift -- I have no idea why the extended linters are canceling. There is no output.

kallewoof committed:
contrib: add signet scripts (issuer, etc)

kallewoof committed:
add witness commitment section support to primitive/block

This also moves GetWitnessCommitmentIndex out of validation.cpp into CBlock, with 3 lines affected (GetWitnessCommitmentIndex(block) -> block.GetWitnessCommitmentIndex()).

kallewoof committed:
add signet basic support (signet.cpp)

kallewoof committed:
qt: update QT to support signet network

kallewoof committed:
consensus: add signet validation

kallewoof committed:
rpc: add signblock command

kallewoof committed:
rpc: include signet commitment in blockToJSON dictionary results

kallewoof committed:
rpc: add getnewblockhex command for obtaining an unsigned block

kallewoof committed:
i/o: make DecodeHexBlk able to read block hex from file

kallewoof committed:
test: add signet mining tests

Also adds a condition to test initialization code to not generate blocks for signet chains.

kallewoof committed:
rpc/signet: show signet blockscript in 'getblockchaininfo'

kallewoof committed:
util_tests.cpp update hash for hard coded chain merge

kallewoof committed:
add signet chain and accompanying parameters

kallewoof committed:
rpc: add grindblock command for generating POW on a signed signet block

kallewoof committed:
rpc: refactor out grindBlock part from generateBlocks

kallewoof committed:
add simple signature support (checker)

kallewoof committed:
add simple signature support (creator)
